File: solutions/fortran/reverse-string/1/reverse_string.f90
module reverse_string
  implicit none
contains

  function reverse(input) result(reversed)
    character(*), intent(in) :: input
    character(len=len(input)) :: reversed
    integer :: i, j, k
    k = len(input)
    do i = 1,k
      j = k - i + 1
      reversed(i:i) = input(j:j)
    end do
  end function

end module

File: solutions/c/reverse-string/1/reverse_string.c
#include <string.h>
#include <stdlib.h>
#include "reverse_string.h"

char *reverse(const char *value)
{
    int len = strlen(value) - 1;
    char* result = malloc(len + 1);

    result[len] = 0;

    if (len == 0)
    {
        return result;
    }

    int last = len;
    int it = 0;
    while (it <= last)
    {
        result[it] = value[last - it];
        ++it;
    }

    return result;
}